<!DOCTYPE html>
<html lang="zh">
  <head>
    
<meta charset="UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=edge" />
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1"/>


<meta http-equiv="Cache-Control" content="no-transform" />
<meta http-equiv="Cache-Control" content="no-siteapp" />



  <meta name="description" content="Exception Control in Python"/>







  <link rel="alternate" href="/atom.xml" title="EOF">




  <link rel="shortcut icon" type="image/x-icon" href="/favicon.ico?v=2.2.x" />



<link rel="canonical" href="http://jasonleaster.github.io/2016/02/22/exception-control-in-python/"/>


<meta name="description" content="Pythoner may be familiar with exception control like the demo beblow there:
123456789try:    i = 0    while True:        i += 1except KeyboardInterrupt:    print &quot;after abort:&quot;, iprint &quot;get here&quot;">
<meta property="og:type" content="article">
<meta property="og:title" content="Exception Control in Python">
<meta property="og:url" content="http://jasonleaster.github.io/2016/02/22/exception-control-in-python/index.html">
<meta property="og:site_name" content="EOF">
<meta property="og:description" content="Pythoner may be familiar with exception control like the demo beblow there:
123456789try:    i = 0    while True:        i += 1except KeyboardInterrupt:    print &quot;after abort:&quot;, iprint &quot;get here&quot;">
<meta property="og:image" content="http://jasonleaster.github.io/images/img_for_2016_02_22/traceback.png">
<meta property="og:image" content="http://jasonleaster.github.io/images/img_for_2016_02_22/frameobjectlist.png">
<meta property="og:image" content="http://jasonleaster.github.io/images/img_for_2016_02_22/hacktime0.png">
<meta property="og:image" content="http://jasonleaster.github.io/images/img_for_2016_02_22/bridge.jpg">
<meta property="og:updated_time" content="2017-02-26T13:28:32.763Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Exception Control in Python">
<meta name="twitter:description" content="Pythoner may be familiar with exception control like the demo beblow there:
123456789try:    i = 0    while True:        i += 1except KeyboardInterrupt:    print &quot;after abort:&quot;, iprint &quot;get here&quot;">
<meta name="twitter:image" content="http://jasonleaster.github.io/images/img_for_2016_02_22/traceback.png">


<link rel="stylesheet" type="text/css" href="/css/style.css?v=2.2.x" />



  <link rel="stylesheet" type="text/css" href="/lib/fancybox/jquery.fancybox.css" />





<script type="text/javascript">
  var themeConfig = {
    search: {
      enable: true,
      path: "/search.xml",
    },
    navbar: {
      enable: true
    },
    fancybox: {
      enable: true
    },
    toc: {
      enable: true
    },
  };
</script>



  



    <title> Exception Control in Python · EOF </title>
  </head>

  <body>
    <div class="container">
      <header id="header" class="header"><div class="logo-wrapper">
  <a href="/." class="logo">EOF</a>
</div>

<nav class="site-nav">
  
    <ul id="menu" class="menu">
      
        <li class="menu-item">
          <a class="menu-item-link" href="/">
            
            
              Home
            
          </a>
        </li>
      
        <li class="menu-item">
          <a class="menu-item-link" href="/archives/">
            
            
              Archives
            
          </a>
        </li>
      
      
        <li class="menu-search">
          <form>
            <i class="iconfont icon-search" id="open-search"></i>
            <input type="text" class="search-input" id="search-input" />
            <i class="iconfont icon-close" id="close-search"></i>
          </form>
        </li>
      
    </ul>
  
</nav>

<div class="mobile-navbar">
  <div class="mobile-header">
    <div class="mobile-header-logo">
      <a href="/." class="logo">EOF</a>
    </div>

    <div class="mobile-header-icon">
      <span></span>
      <span></span>
      <span></span>
    </div>
  </div>
  <nav class="mobile-menu">
    
      <a class="mobile-menu-item" href="/">
        
        
          Home
        
      </a>
    
      <a class="mobile-menu-item" href="/archives/">
        
        
          Archives
        
      </a>
    
  </nav>
</div>
      </header>

      <main id="main" class="main">
        <div class="content-wrapper">
          <div id="content" class="content">
            
  
  <article class="post">
    <header class="post-header">
      <h1 class="post-title">
        
          Exception Control in Python
        
      </h1>

      <div class="post-meta">
        <span class="post-time">
          Feb 22, 2016
        </span>
      </div>
    </header>

    
    
  <div class="post-toc" id="post-toc">
    <h2 class="post-toc-title">Contents</h2>
    <div class="post-toc-content">
      <ol class="toc"><li class="toc-item toc-level-3"><a class="toc-link" href="#Hacker-Time"><span class="toc-text">Hacker Time</span></a></li></ol>
    </div>
  </div>


    <div class="post-content">
      
        <p>Pythoner may be familiar with exception control like the demo beblow there:</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">try</span>:</div><div class="line">    i = <span class="number">0</span></div><div class="line">    <span class="keyword">while</span> <span class="keyword">True</span>:</div><div class="line">        i += <span class="number">1</span></div><div class="line"></div><div class="line"><span class="keyword">except</span> KeyboardInterrupt:</div><div class="line">    <span class="keyword">print</span> <span class="string">"after abort:"</span>, i</div><div class="line"></div><div class="line"><span class="keyword">print</span> <span class="string">"get here"</span></div></pre></td></tr></table></figure>
<a id="more"></a>
<p>If you press down <code>ctrl + c</code>, there will trigger a interruption by keyboard. The inner infinite loop will stop and jump into the exception handler. There are another examples of exception and interruption in the computer. Programmer may also ask what’s the benefite to handle the exception …</p>
<p>What if there is something unpredictable and it will stop the program to run correctly?</p>
<p>If you do 1 divide 0 in your program, the CPU don’t know how to compute that expression. In the level of operating system, OS will handle it as an exception and tell programmer that it doesn’t work. The program must be stopped or killed.</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">try</span>:</div><div class="line">    <span class="number">1</span>/<span class="number">0</span></div><div class="line"><span class="keyword">except</span>:</div><div class="line">    <span class="keyword">print</span> <span class="string">"Welcome Chapter 42 -- Guideline to galaxy"</span></div></pre></td></tr></table></figure>
<p>Although there trig a exception for diveding zero, but we catch it and handle it correctly. So, the program end correctly.</p>
<p>The mechanism of exception handling come from operating system. </p>
<p>Here is the definition of exception object in Python. (Don’t forget that everything in Python is object).</p>
<figure class="highlight c"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div></pre></td><td class="code"><pre><div class="line">[Include/pyerrors.h]</div><div class="line"></div><div class="line">...</div><div class="line"></div><div class="line"><span class="keyword">typedef</span> <span class="keyword">struct</span> &#123;</div><div class="line">    PyObject_HEAD</div><div class="line">    PyObject *dict;</div><div class="line">    PyObject *args;</div><div class="line">    PyObject *message;</div><div class="line">&#125; PyBaseExceptionObject;</div><div class="line"></div><div class="line">...</div><div class="line"></div><div class="line"><span class="comment">/* Predefined exceptions */</span></div><div class="line"></div><div class="line">PyAPI_DATA(PyObject *) PyExc_BaseException;</div><div class="line">PyAPI_DATA(PyObject *) PyExc_Exception;</div><div class="line">PyAPI_DATA(PyObject *) PyExc_StopIteration;</div><div class="line">...</div><div class="line">PyAPI_DATA(PyObject *) PyExc_ZeroDivisionError;</div><div class="line">PyAPI_DATA(PyObject *) PyExc_EOFError;</div><div class="line"></div><div class="line">...</div></pre></td></tr></table></figure>
<p>Here is a example about how python deal with the exception.</p>
<p>What would happen if there is a expression <code>1/0</code>.<br>I use IPython interpreter do this demo.</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div></pre></td><td class="code"><pre><div class="line">In [<span class="number">18</span>]: code = compile(<span class="string">"1/0"</span>, <span class="string">"testscript"</span>, mode= <span class="string">"exec"</span>)</div><div class="line"></div><div class="line">In [<span class="number">19</span>]: dis.disassemble(code)</div><div class="line">      <span class="number">1</span>           <span class="number">0</span> LOAD_CONST               <span class="number">0</span> (<span class="number">1</span>)</div><div class="line">                  <span class="number">3</span> LOAD_CONST               <span class="number">1</span> (<span class="number">0</span>)</div><div class="line">                  <span class="number">6</span> BINARY_DIVIDE       </div><div class="line">                  <span class="number">7</span> POP_TOP             </div><div class="line">                  <span class="number">8</span> LOAD_CONST               <span class="number">2</span> (<span class="keyword">None</span>)</div><div class="line">                 <span class="number">11</span> RETURN_VALUE</div></pre></td></tr></table></figure>
<p>The assemble code (opcode) of expression <code>1/0</code> in Python is compiled into that opcode code. It isn’t difficult to understand what the <code>LOAD_CONST</code> do.</p>
<p>Let’s dig into the detail of <code>BINARY_DIVIDE</code> in Python/ceval.c</p>
<figure class="highlight c"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div><div class="line">32</div><div class="line">33</div><div class="line">34</div><div class="line">35</div><div class="line">36</div><div class="line">37</div><div class="line">38</div><div class="line">39</div><div class="line">40</div><div class="line">41</div><div class="line">42</div><div class="line">43</div><div class="line">44</div><div class="line">45</div><div class="line">46</div><div class="line">47</div><div class="line">48</div><div class="line">49</div><div class="line">50</div><div class="line">51</div><div class="line">52</div><div class="line">53</div><div class="line">54</div><div class="line">55</div><div class="line">56</div><div class="line">57</div><div class="line">58</div><div class="line">59</div><div class="line">60</div><div class="line">61</div><div class="line">62</div><div class="line">63</div><div class="line">64</div><div class="line">65</div><div class="line">66</div><div class="line">67</div><div class="line">68</div><div class="line">69</div><div class="line">70</div><div class="line">71</div><div class="line">72</div><div class="line">73</div><div class="line">74</div><div class="line">75</div><div class="line">76</div><div class="line">77</div><div class="line">78</div><div class="line">79</div><div class="line">80</div></pre></td><td class="code"><pre><div class="line">[Python/ceval.c]</div><div class="line">    <span class="keyword">for</span>(;;)<span class="comment">// big for loop</span></div><div class="line">        ...</div><div class="line">        TARGET_NOARG(BINARY_DIVIDE)</div><div class="line">            &#123;</div><div class="line">                <span class="keyword">if</span> (!_Py_QnewFlag) &#123;</div><div class="line">                    w = POP();</div><div class="line">                    v = TOP();</div><div class="line">                    x = PyNumber_Divide(v, w); <span class="comment">//</span></div><div class="line">                    Py_DECREF(v);</div><div class="line">                    Py_DECREF(w);</div><div class="line">                    SET_TOP(x);</div><div class="line">                    <span class="keyword">if</span> (x != <span class="literal">NULL</span>) DISPATCH();</div><div class="line">                    <span class="keyword">break</span>;</div><div class="line">                &#125;</div><div class="line">            &#125;</div><div class="line">        ...</div><div class="line"></div><div class="line">        <span class="comment">/* </span></div><div class="line">           set up the basic information for the reason </div><div class="line">           why exception happened  -- notes by Jason Leaster</div><div class="line">         */</div><div class="line">        <span class="comment">/* Quickly continue if no error occurred */</span></div><div class="line"></div><div class="line">        <span class="keyword">if</span> (why == WHY_NOT) &#123;</div><div class="line">            <span class="keyword">if</span> (err == <span class="number">0</span> &amp;&amp; x != <span class="literal">NULL</span>) &#123;</div><div class="line">                READ_TIMESTAMP(loop1);</div><div class="line">                <span class="keyword">continue</span>; <span class="comment">/* Normal, fast path */</span></div><div class="line">            &#125;</div><div class="line">            why = WHY_EXCEPTION;</div><div class="line">            x = Py_None;</div><div class="line">            err = <span class="number">0</span>;</div><div class="line">        &#125;</div><div class="line"></div><div class="line">        ...</div><div class="line"></div><div class="line">        <span class="comment">/* Log traceback info if this is a real exception */</span></div><div class="line">        <span class="keyword">if</span> (why == WHY_EXCEPTION) &#123;</div><div class="line">            PyTraceBack_Here(f);</div><div class="line">            ...</div><div class="line">        &#125;</div><div class="line">        ...</div><div class="line"></div><div class="line">[Objects/abstract.c]</div><div class="line">BINARY_FUNC(PyNumber_Subtract, nb_subtract, <span class="string">"-"</span>)</div><div class="line"></div><div class="line"><span class="comment">/* PyNumber_Divide is just the other name of function nb_divide </span></div><div class="line">   and then you will find that the nb_divide is just a function pointer</div><div class="line">   in struct PyNumberMethods.</div><div class="line"></div><div class="line">   The REAL implementation is function @int_classic_div in </div><div class="line">   Objects/intobject.c. The @int_classic_div will call function</div><div class="line">   @i_divmod in Objects/intobject.c</div><div class="line"></div><div class="line">                                    -- notes by Jason Leaster</div><div class="line"> */</div><div class="line">BINARY_FUNC(PyNumber_Divide, nb_divide, <span class="string">"/"</span>) </div><div class="line"></div><div class="line">BINARY_FUNC(PyNumber_Divmod, nb_divmod, <span class="string">"divmod()"</span>)</div><div class="line"></div><div class="line">[Objects/intobject.c]</div><div class="line">i_divmod(<span class="keyword">register</span> <span class="keyword">long</span> x, <span class="keyword">register</span> <span class="keyword">long</span> y,</div><div class="line">           <span class="keyword">long</span> *p_xdivy, <span class="keyword">long</span> *p_xmody)</div><div class="line">&#123;</div><div class="line">    <span class="keyword">long</span> xdivy, xmody;</div><div class="line"></div><div class="line">    <span class="keyword">if</span> (y == <span class="number">0</span>) &#123;</div><div class="line">        <span class="comment">/*</span></div><div class="line">         Keypoint!</div><div class="line">         Here is where Python setup a new exception for "divide zero"</div><div class="line">         problem.</div><div class="line">                                            -- notes by JasonLeaster</div><div class="line">         */</div><div class="line">        PyErr_SetString(PyExc_ZeroDivisionError,</div><div class="line">                <span class="string">"integer division or modulo by zero"</span>);</div><div class="line">        <span class="keyword">return</span> DIVMOD_ERROR;</div><div class="line">    &#125;</div><div class="line"></div><div class="line">    ... ...</div><div class="line">&#125;</div></pre></td></tr></table></figure>
<p>It’s the last mile to get the target place.<br><code>PyErr_SetString</code> finish the job about the initialization of <code>divide zero</code> exception.</p>
<figure class="highlight c"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div><div class="line">32</div><div class="line">33</div><div class="line">34</div><div class="line">35</div><div class="line">36</div><div class="line">37</div><div class="line">38</div></pre></td><td class="code"><pre><div class="line">[Python/errors.c]</div><div class="line"></div><div class="line"><span class="function"><span class="keyword">void</span></span></div><div class="line"><span class="title">PyErr_Restore</span><span class="params">(PyObject *type, PyObject *value, PyObject *traceback)</span></div><div class="line">&#123;</div><div class="line">    PyThreadState *tstate = PyThreadState_GET();</div><div class="line">    PyObject *oldtype, *oldvalue, *oldtraceback;</div><div class="line"></div><div class="line">    ... </div><div class="line"></div><div class="line">    <span class="comment">/* Save these in locals to safeguard against recursive</span></div><div class="line">       invocation through Py_XDECREF */</div><div class="line">    oldtype = tstate-&gt;curexc_type;</div><div class="line">    oldvalue = tstate-&gt;curexc_value;</div><div class="line">    oldtraceback = tstate-&gt;curexc_traceback;</div><div class="line"></div><div class="line">    tstate-&gt;curexc_type = type;</div><div class="line">    tstate-&gt;curexc_value = value;</div><div class="line">    tstate-&gt;curexc_traceback = traceback;</div><div class="line">    ...</div><div class="line">&#125;</div><div class="line"></div><div class="line"><span class="function"><span class="keyword">void</span></span></div><div class="line"><span class="title">PyErr_SetObject</span><span class="params">(PyObject *exception, PyObject *value)</span></div><div class="line">&#123;</div><div class="line">    ...</div><div class="line"></div><div class="line">    <span class="comment">/* log this exception message into thread state */</span></div><div class="line">    PyErr_Restore(exception, value, (PyObject *)<span class="literal">NULL</span>);</div><div class="line">&#125;</div><div class="line"></div><div class="line"><span class="function"><span class="keyword">void</span></span></div><div class="line"><span class="title">PyErr_SetString</span><span class="params">(PyObject *exception, <span class="keyword">const</span> <span class="keyword">char</span> *<span class="built_in">string</span>)</span></div><div class="line">&#123;</div><div class="line">    PyObject *value = PyString_FromString(<span class="built_in">string</span>);</div><div class="line">    PyErr_SetObject(exception, value);</div><div class="line">    Py_XDECREF(value);</div><div class="line">&#125;</div></pre></td></tr></table></figure>
<p>After the VM (ceval.c) know why there is a exception, the VM will take step<br>into handling this exception. We have knew that the big <code>for switch loop</code> execute the opcode step by step.<br>Once there is something wrong with the opcode at runtime, the VM will setup<br> exception object by <code>PyErr_SetObject</code>. And then VM also will setup the traceback object of type <code>PyTracebackObject</code>.</p>
<figure class="highlight c"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div></pre></td><td class="code"><pre><div class="line"><span class="comment">/* Traceback interface */</span></div><div class="line"></div><div class="line"><span class="keyword">typedef</span> <span class="keyword">struct</span> _traceback &#123;</div><div class="line">    PyObject_HEAD</div><div class="line">    <span class="keyword">struct</span> _traceback *tb_next;</div><div class="line">    <span class="keyword">struct</span> _frame *tb_frame;</div><div class="line">    <span class="keyword">int</span> tb_lasti;</div><div class="line">    <span class="keyword">int</span> tb_lineno;</div><div class="line">&#125; PyTracebackObject;</div></pre></td></tr></table></figure>
<p>You may notice that the <code>PyTracebackObject</code> is a single direction linked-list. </p>
<figure class="highlight c"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div><div class="line">19</div><div class="line">20</div><div class="line">21</div><div class="line">22</div><div class="line">23</div><div class="line">24</div><div class="line">25</div><div class="line">26</div><div class="line">27</div><div class="line">28</div><div class="line">29</div><div class="line">30</div><div class="line">31</div><div class="line">32</div><div class="line">33</div><div class="line">34</div></pre></td><td class="code"><pre><div class="line"><span class="function"><span class="keyword">static</span> PyTracebackObject *</span></div><div class="line"><span class="title">newtracebackobject</span><span class="params">(PyTracebackObject *next, PyFrameObject *frame)</span></div><div class="line">&#123;</div><div class="line">    PyTracebackObject *tb;</div><div class="line">    <span class="keyword">if</span> ((next != <span class="literal">NULL</span> &amp;&amp; !PyTraceBack_Check(next)) ||</div><div class="line">            frame == <span class="literal">NULL</span> || !PyFrame_Check(frame)) &#123;</div><div class="line">        PyErr_BadInternalCall();</div><div class="line">        <span class="keyword">return</span> <span class="literal">NULL</span>;</div><div class="line">    &#125;</div><div class="line">    tb = PyObject_GC_New(PyTracebackObject, &amp;PyTraceBack_Type);</div><div class="line">    <span class="keyword">if</span> (tb != <span class="literal">NULL</span>) &#123;</div><div class="line">        Py_XINCREF(next);</div><div class="line">        tb-&gt;tb_next = next;</div><div class="line">        Py_XINCREF(frame);</div><div class="line">        tb-&gt;tb_frame = frame;</div><div class="line">        tb-&gt;tb_lasti = frame-&gt;f_lasti;</div><div class="line">        tb-&gt;tb_lineno = PyFrame_GetLineNumber(frame);</div><div class="line">        PyObject_GC_Track(tb);</div><div class="line">    &#125;</div><div class="line">    <span class="keyword">return</span> tb;</div><div class="line">&#125;</div><div class="line"></div><div class="line"><span class="function"><span class="keyword">int</span></span></div><div class="line"><span class="title">PyTraceBack_Here</span><span class="params">(PyFrameObject *frame)</span></div><div class="line">&#123;</div><div class="line">    PyThreadState *tstate = PyThreadState_GET();</div><div class="line">    PyTracebackObject *oldtb = (PyTracebackObject *) tstate-&gt;curexc_traceback;</div><div class="line">    PyTracebackObject *tb = newtracebackobject(oldtb, frame);</div><div class="line">    <span class="keyword">if</span> (tb == <span class="literal">NULL</span>)</div><div class="line">        <span class="keyword">return</span> <span class="number">-1</span>;</div><div class="line">    tstate-&gt;curexc_traceback = (PyObject *)tb;</div><div class="line">    Py_XDECREF(oldtb);</div><div class="line">    <span class="keyword">return</span> <span class="number">0</span>;</div><div class="line">&#125;</div></pre></td></tr></table></figure>
<figure class="highlight python"><table><tr><td class="gutter"><pre><div class="line">1</div><div class="line">2</div><div class="line">3</div><div class="line">4</div><div class="line">5</div><div class="line">6</div><div class="line">7</div><div class="line">8</div><div class="line">9</div><div class="line">10</div><div class="line">11</div><div class="line">12</div><div class="line">13</div><div class="line">14</div><div class="line">15</div><div class="line">16</div><div class="line">17</div><div class="line">18</div></pre></td><td class="code"><pre><div class="line"><span class="keyword">import</span> sys</div><div class="line"></div><div class="line"><span class="function"><span class="keyword">def</span> <span class="title">h</span><span class="params">()</span>:</span></div><div class="line">    <span class="keyword">print</span> <span class="string">"in frame :"</span>, sys._getframe()</div><div class="line">    <span class="keyword">print</span> <span class="string">"in function :"</span>, sys._getframe().f_code.co_name</div><div class="line">    <span class="number">1</span>/<span class="number">0</span></div><div class="line"></div><div class="line"><span class="function"><span class="keyword">def</span> <span class="title">g</span><span class="params">()</span>:</span></div><div class="line">    <span class="keyword">print</span> <span class="string">"in frame :"</span>, sys._getframe()</div><div class="line">    <span class="keyword">print</span> <span class="string">"in function :"</span>, sys._getframe().f_code.co_name</div><div class="line">    h()</div><div class="line"></div><div class="line"><span class="function"><span class="keyword">def</span> <span class="title">f</span><span class="params">()</span>:</span></div><div class="line">    <span class="keyword">print</span> <span class="string">"in frame :"</span>, sys._getframe()</div><div class="line">    <span class="keyword">print</span> <span class="string">"in function :"</span>, sys._getframe().f_code.co_name</div><div class="line">    g()</div><div class="line"></div><div class="line">f()</div></pre></td></tr></table></figure>
<p>This program will trig the exception at function h().<br>Here is the output of that program.</p>
<p><img src="/images/img_for_2016_02_22/traceback.png" alt="images"></p>
<p><img src="/images/img_for_2016_02_22/frameobjectlist.png" alt="images"></p>
<h3 id="Hacker-Time"><a href="#Hacker-Time" class="headerlink" title="Hacker Time"></a>Hacker Time</h3><p><img src="/images/img_for_2016_02_22/hacktime0.png" alt="images"></p>
<hr>
<p>Photo by Jason Leaster, ChangeDe, HuNan, China<br><img src="/images/img_for_2016_02_22/bridge.jpg" alt="images"></p>

      
    </div>

    
      
      

  <div class="post-copyright">
    <p class="copyright-item">
      <span>作者: </span>
      <span>Jason Leaster</span>
    </p>
    <p class="copyright-item">
      <span>来源: </span>
      <a href="http://jasonleaster.github.io">http://jasonleaster.github.io</a>
    </p>
    <p class="copyright-item">
      <span>链接: </span>
      <a href="http://jasonleaster.github.io/2016/02/22/exception-control-in-python/">http://jasonleaster.github.io/2016/02/22/exception-control-in-python/</a>
    </p>

    <p class="copyright-item lincese">
      
      本文采用<a rel="license" href="http://creativecommons.org/licenses/by-nc/4.0/" target="_blank">知识共享署名-非商业性使用 4.0 国际许可协议</a>进行许可
    </p>
  </div>



      
      
    

    
  </article>


          </div>
          
  <div class="comments" id="comments">
    
  </div>

        </div>  
      </main>

      <footer id="footer" class="footer">

  <div class="social-links">
    
      
        
          <a href="mailto:jasonleaster@gmail.com" class="iconfont icon-email" title="email"></a>
        
      
    
      
    
      
    
      
    
      
    
      
    
      
        
          <a href="https://github.com/jasonleaster" class="iconfont icon-github" title="github"></a>
        
      
    
      
    
      
        
          <a href="https://zhihu.com/jasonleaster" class="iconfont icon-zhihu" title="zhihu"></a>
        
      
    
    
    
  </div>


<div class="copyright">
  <span class="power-by">
    Powered by <a class="hexo-link" href="https://hexo.io/">Hexo</a>
  </span>
  
  <span class="theme-info">
    Theme - 
    <a class="theme-link" href="https://github.com/ahonn/hexo-theme-even">Even</a>
  </span>

  <span class="copyright-year">
    
    &copy; 
     
      2015 - 
    
    2018

    <span class="heart">
      <i class="iconfont icon-heart"></i>
    </span>
    <span class="author">Jason Leaster</span>
  </span>
</div>
      </footer>

      <div class="back-to-top" id="back-to-top">
        <i class="iconfont icon-up"></i>
      </div>
    </div>

    


    
  





  
    <script type="text/javascript" src="/lib/jquery/jquery-3.1.1.min.js"></script>
  

  
    <script type="text/javascript" src="/lib/fancybox/jquery.fancybox.pack.js"></script>
  

    <script type="text/javascript" src="/js/src/even.js?v=2.2.x"></script>
<script type="text/javascript" src="/js/src/bootstrap.js?v=2.2.x"></script>

  </body>
</html>